### PR TITLE
refactor:adjust special cases

### LINKED ISSUE #22: ZPCA output: special cases
**Is your feature request related to a problem? Please describe.**
It turns out that [ZPCA](https://github.com/zavolanlab/zpca) output tables look differently for 2+ samples and for 2 samples only. Also: if there is just one sample in our bigger pipeline (zarp) MultiQC will still run at the end but there will be no table for PCA analysis - ZPCA will just create a mock empty file. We need to account for these two additional cases in our parsing module here.

**Describe the solution you'd like**
1. Plotting plugin should properly parse output tables after running ZPCA on 2+ samples as well as 2 samples.
(The output table differs just by one column, but still - this is enough for a crash at the current moment)
2. Plotting plugin should detect that the ZPCA table is an empty file and in such case skip plotting at all, if It is possible? It would be even better if we could design the plugin in such a way that in this special case the whole "ZPCA" section is removed from the final report (so that we do not have an isolated title without any plot inside it - that would look weird...)

**Additional context**
I asked @fgypas to add unit tests to his repo for these two special cases. So that now once we run `test.py` for his tool it creates directories for all three test cases. We should use the output of these tests as input, based on which the plugin has to be adjusted.
Once the job is done it would be good to follow the same logic n **this** repository: under `tests/zpca/` we should have three distinct sub-directories with these three distinct input types and three example html reports.

@krish8484 - could you please look into this?

### PR DESCRIPTION
**Pull request by krish8484**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #22 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

### CODE REVIEW

**Comment by krish8484** [comment]
Hi @AngryMaciek @fgypas, I have updated the code to handle special cases, please have a look.

File: modules/zpca/zpca.py [hunk 1]
```diff
@@ -146,20 +203,31 @@ def parse_zpca_logs(self, f: dict):
...
-        # Concatinating each character
+        # Concatinating each word
         for k in words:
-            listToStr1.append("".join([str(elem) for elem in k]))
-        # Conactinating each word
-        for k in listToStr1:
             listToStr.append("".join([str(elem) for elem in k]))
 
+        len_listToStr = len(listToStr)
+        if len_listToStr == 36:
```
Comment (fgypas) [inline modules/zpca/zpca.py]: I am not sure what is happening here, but is it safe to check based on the length of the list?

**Review by fgypas — COMMENTED** [review]
LGTM. I am not only sure about one check you are doing based on the length, but I think that it is fine. @AngryMaciek Can you please also have a look. I am not so familiar with multiqc.

File: .flake8 [hunk 2]
```diff
@@ -1,3 +1,3 @@
 [flake8]
-max-line-length = 88
+max-line-length = 90
```
Comment (AngryMaciek) [inline .flake8:2]: Generally people should stick to the rules of formatting for a given project. I am not against this change (since `88` is arbitrary anyway) but I would like to know why did you need to introduce this modification?

File: modules/zpca/zpca.py [hunk 3]
```diff
@@ -111,22 +150,42 @@ def parse_scree_logs(self, f: dict):
...
 
-        exp_car_str = "Percentage of Explained Variance"
+        len_listToStr = len(listToStr)
+
+        if len_listToStr == 8:
+            return self._parse_scree_helper_3(listToStr)
+        elif len_listToStr == 6:
+            return self._parse_scree_helper_2(listToStr)
+        else:
+            return {}
```
Comment (AngryMaciek) [inline modules/zpca/zpca.py]: I generally do not understand the parsing logic with these `words` vs. `characters`, `listToStr` len being `6` or `8`. Could you please add a little explanation to the code (in a comment)?

File: modules/zpca/zpca.py (hunk 1 above)
Comment (AngryMaciek) [reply modules/zpca/zpca.py]: I second that; this code fragment is similar to the one above for which I also raised concern. Could you please add some info in the comment here as well?

**Review by AngryMaciek — COMMENTED** [review]
Look great! The _HTML_ reports are exactly as I imagined them, awesome.
I would just like to see a little clarification to your actions.

@krish8484

File: .flake8 (hunk 2 above)
Comment (krish8484) [reply .flake8:2]: In our project, we have 2 formatting checks i.e `black` and `flake8` (even though I think either of them is enough), after formatting the code using `black` it allows lines upto 90 characters which is not the case with `flake8`. They both conflict, hence this change.

File: modules/zpca/zpca.py (hunk 1 above)
Comment (krish8484) [reply modules/zpca/zpca.py]: Comments added!

Comment (AngryMaciek) [reply modules/zpca/zpca.py]: @krish8484 
These comments are not very descriptive... Could you please explain (in the comments) how does `len_listToStr` being `6` `8` or `36` correspond to the number of principal components: `1`, `2` `3`?
What is the logic and intention here? what do you check? why these exact numbers appear?

**Comment by krish8484** [comment]
Now when I look back, writing the same thing this (commit 403afb4) way makes it more appropriate and less complicated.

**Comment by AngryMaciek** [comment]
Ah yes, now I understand it. It is generally more "Pythonic" to operate on names and use build-in methods, whereas indices are more C-like.

Thanks! I will merge now, push a new container and we'll see if it integrates well with the pipeline.
